radix-engine/src/track/track.rs

```diff
@@ -15,40 +19,193 @@ impl SubstateLockState {
     pub fn no_lock() -> Self {
         Self::Read(0)
     }
+
+    pub fn is_locked(&self) -> bool {
+        !matches!(self, SubstateLockState::Read(0usize))
+    }
+
+    pub fn try_lock(&mut self, flags: LockFlags) -> Result<(), SubstateLockError> {
+        match self {
+            SubstateLockState::Read(n) => {
+                if flags.contains(LockFlags::MUTABLE) {
+                    if *n != 0 {
+                        return Err(SubstateLockError);
+                    }
+                    *self = SubstateLockState::Write;
+                } else {
+                    *n = *n + 1;
+                }
+            }
+            SubstateLockState::Write => {
+                return Err(SubstateLockError);
+            }
+        }
+
+        Ok(())
+    }
+
+    fn unlock(&mut self) {
+        match self {
+            SubstateLockState::Read(n) => {
+                *n = *n - 1;
+            }
+            SubstateLockState::Write => {
+                *self = SubstateLockState::no_lock();
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct RuntimeSubstate {
+    pub value: IndexedScryptoValue,
+    lock_state: SubstateLockState,
+}
+
+impl RuntimeSubstate {
+    fn new(value: IndexedScryptoValue) -> Self {
+        Self {
+            value,
+            lock_state: SubstateLockState::no_lock(),
+        }
+    }
 }
 
-#[derive(Debug)]
-pub enum ExistingMetaState {
-    Loaded,
-    Updated(Option<IndexedScryptoValue>),
+#[derive(Clone, Debug)]
+pub enum ReadOnly {
+    NonExistent,
+    Existent(RuntimeSubstate),
 }
 
-#[derive(Debug)]
-pub enum SubstateMetaState {
-    New,
-    Existing { state: ExistingMetaState },
+#[derive(Clone, Debug)]
+pub enum Read {
+    NonExistent,
+    Existent,
 }
 
-#[derive(Debug)]
-pub struct LoadedSubstate {
-    substate: IndexedScryptoValue,
-    lock_state: SubstateLockState,
-    meta_state: SubstateMetaState,
+#[derive(Clone, Debug)]
+pub enum Write {
+    Update(RuntimeSubstate),
+    Delete,
+}
+
+impl Write {
+    fn into_value(self) -> Option<IndexedScryptoValue> {
+        match self {
+            Write::Update(substate) => Some(substate.value),
+            Write::Delete => None,
+        }
+    }
+}
+
+// TODO: Add new virtualized
+#[derive(Clone, Debug)]
+pub enum TrackedSubstateKey {
+    New(RuntimeSubstate),
+    ReadOnly(ReadOnly),
+    ReadAndWrite(Read, Write),
```

Comment:
`Read` for `ReadAndWrite` should not have a `RuntimeSubstate` since a lock isn't needed. But to keep track of the original read is implemented in the next PR.